File: androidApp/src/main/kotlin/com/virginactive/android/ui/detail/ClassDetailViewModel.kt
package com.virginactive.android.ui.detail

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.virginactive.shared.domain.booking.BookClassUseCase
import com.virginactive.shared.domain.booking.BookingOutcome
import com.virginactive.shared.domain.booking.CancelBookingUseCase
import com.virginactive.shared.domain.booking.ClassInstance
import com.virginactive.shared.domain.booking.ForfeitPolicy
import com.virginactive.shared.domain.booking.UserBookingStatus
import com.virginactive.shared.domain.error.AppResult
import com.virginactive.shared.domain.error.DomainError
import com.virginactive.shared.domain.profile.GetProfileUseCase
import com.virginactive.shared.domain.timetable.GetTimetableUseCase
import kotlinx.coroutines.launch
import java.util.Calendar

class ClassDetailViewModel(
    private val getProfile: GetProfileUseCase,
    private val getTimetable: GetTimetableUseCase,
    private val bookClass: BookClassUseCase,
    private val cancelBooking: CancelBookingUseCase,
) : ViewModel() {

    var state by mutableStateOf(ClassDetailUiState())
        private set

    fun load(classId: String) {
        if (state.isLoading) return
        state = state.copy(isLoading = true, error = null)
        viewModelScope.launch {
            val clubId = when (val profile = getProfile()) {
                is AppResult.Ok -> profile.value.homeClub.id
                is AppResult.Err -> ""
            }
            when (val week = getTimetable(clubId, today())) {
                is AppResult.Ok -> {
                    val instance = week.value.days
                        .flatMap { it.classes }
                        .firstOrNull { it.classId == classId }
                    if (instance == null) {
                        state = state.copy(
                            isLoading = false,
                            error = DomainError.NotFound("ClassNotFound"),
                            clubId = clubId,
                        )
                    } else {
                        state = state.copy(
                            isLoading = false,
                            classInstance = instance,
                            clubId = clubId,
                            confirmation = Confirmation(
                                bookingId = null,
                                userBookingStatus = instance.userBookingStatus,
                                waitlistPosition = null,
                            ),
                        )
                    }
                }

                is AppResult.Err -> state = state.copy(isLoading = false, error = week.error)
            }
        }
    }

    fun book() {
        val instance = state.classInstance ?: return
        val clubId = state.clubId ?: return
        if (state.actionInFlight) return
        state = state.copy(actionInFlight = true, actionError = null)
        viewModelScope.launch {
            when (val result = bookClass(clubId, instance.classId)) {
                is AppResult.Ok -> state = state.copy(
                    actionInFlight = false,
                    confirmation = result.value.toConfirmation(),
                )

                is AppResult.Err -> state = state.copy(
                    actionInFlight = false,
                    actionError = result.error,
                )
            }
        }
    }

    fun cancel() {
        val instance = state.classInstance ?: return
        val clubId = state.clubId ?: return
        if (state.actionInFlight) return
        state = state.copy(actionInFlight = true, actionError = null)
        viewModelScope.launch {
            when (val result = cancelBooking(clubId, instance.classId)) {
                is AppResult.Ok -> state = state.copy(
                    actionInFlight = false,
                    confirmation = Confirmation(
                        bookingId = null,
                        userBookingStatus = UserBookingStatus.NONE,
                        waitlistPosition = null,
                    ),
                )

                is AppResult.Err -> state = state.copy(
                    actionInFlight = false,
                    actionError = result.error,
                )
            }
        }
    }

    fun actionErrorShown() {
        state = state.copy(actionError = null)
    }

    fun isWithinForfeitWindow(): Boolean {
        val startsAt = state.classInstance?.startsAt ?: return false
        return ForfeitPolicy.isWithinForfeitWindow(startsAt)
    }

    private fun today(): String {
        val now = Calendar.getInstance()
        val year = now.get(Calendar.YEAR)
        val month = now.get(Calendar.MONTH) + 1
        val day = now.get(Calendar.DAY_OF_MONTH)
        return buildString {
            append(year.toString().padStart(4, '0'))
            append('-')
            append(month.toString().padStart(2, '0'))
            append('-')
            append(day.toString().padStart(2, '0'))
        }
    }
}

private fun BookingOutcome.toConfirmation(): Confirmation = when (this) {
    is BookingOutcome.Confirmed -> Confirmation(
        bookingId = booking.bookingId,
        userBookingStatus = UserBookingStatus.BOOKED,
        waitlistPosition = null,
    )

    is BookingOutcome.Waitlisted -> Confirmation(
        bookingId = booking.bookingId,
        userBookingStatus = UserBookingStatus.WAITLISTED,
        waitlistPosition = position,
    )

    is BookingOutcome.AlreadyBooked -> Confirmation(
        bookingId = null,
        userBookingStatus = if (wasWaitlist) UserBookingStatus.WAITLISTED else UserBookingStatus.BOOKED,
        waitlistPosition = null,
    )
}

data class Confirmation(
    val bookingId: String?,
    val userBookingStatus: UserBookingStatus,
    val waitlistPosition: Int?,
)

data class ClassDetailUiState(
    val isLoading: Boolean = false,
    val error: DomainError? = null,
    val classInstance: ClassInstance? = null,
    val clubId: String? = null,
    val confirmation: Confirmation? = null,
    val actionInFlight: Boolean = false,
    val actionError: DomainError? = null,
)

File: iosApp/iosApp/iosApp/ClassDetail/ClassDetailViewModel.swift
import Foundation
import Combine
import Shared

struct Confirmation {
    let bookingId: String?
    let userBookingStatus: UserBookingStatus
    let waitlistPosition: Int?
}

@MainActor
final class ClassDetailViewModel: ObservableObject {
    @Published var isLoading: Bool = false
    @Published var error: DomainError?
    @Published var classInstance: ClassInstance?
    @Published var clubId: String?
    @Published var confirmation: Confirmation?
    @Published var actionInFlight: Bool = false
    @Published var actionError: DomainError?
    @Published var actionErrorMessage: String?

    private let getProfileUseCase: GetProfileUseCase
    private let getTimetableUseCase: GetTimetableUseCase
    private let bookClassUseCase: BookClassUseCase
    private let cancelBookingUseCase: CancelBookingUseCase

    private var loadTask: Task<Void, Never>?
    private var actionTask: Task<Void, Never>?

    init(component: AppComponent) {
        self.getProfileUseCase = component.getProfileUseCase
        self.getTimetableUseCase = component.getTimetableUseCase
        self.bookClassUseCase = component.bookClassUseCase
        self.cancelBookingUseCase = component.cancelBookingUseCase
    }


    func load(classId: String) {
        guard !isLoading else { return }
        isLoading = true
        error = nil
        loadTask = Task {
            let resolvedClubId: String
            switch onEnum(of: try? await getProfileUseCase.invoke()) {
            case .ok(let ok)?:
                resolvedClubId = (ok.value as? UserProfile)?.homeClub.id ?? ""
            default:
                resolvedClubId = ""
            }

            guard !Task.isCancelled else { return }

            switch onEnum(of: try? await getTimetableUseCase.invoke(clubId: resolvedClubId, date: Self.today())) {
            case .ok(let ok)?:
                let week = ok.value as? WeeklyTimetable
                let instance = week?.days
                    .flatMap { $0.classes }
                    .first { $0.classId == classId }
                if let instance {
                    classInstance = instance
                    clubId = resolvedClubId
                    confirmation = Confirmation(
                        bookingId: nil,
                        userBookingStatus: instance.userBookingStatus,
                        waitlistPosition: nil
                    )
                    isLoading = false
                } else {
                    error = DomainErrorNotFound(code: "ClassNotFound")
                    clubId = resolvedClubId
                    isLoading = false
                }
            case .err(let err)?:
                error = err.error
                isLoading = false
            case .none:
                isLoading = false
            }
        }
    }


    func book() {
        guard let instance = classInstance, let clubId, !actionInFlight else { return }
        actionInFlight = true
        actionError = nil
        actionTask = Task {
            switch onEnum(of: try? await bookClassUseCase.invoke(clubId: clubId, classId: instance.classId)) {
            case .ok(let ok)?:
                if let outcome = ok.value as? BookingOutcome {
                    confirmation = Self.toConfirmation(outcome)
                }
                actionInFlight = false
            case .err(let err)?:
                actionError = err.error
                actionErrorMessage = "We couldn't complete that. Tap to try again."
                actionInFlight = false
            case .none:
                actionInFlight = false
            }
        }
    }


    func cancel() {
        guard let instance = classInstance, let clubId, !actionInFlight else { return }
        actionInFlight = true
        actionError = nil
        actionTask = Task {
            switch onEnum(of: try? await cancelBookingUseCase.invoke(clubId: clubId, classId: instance.classId)) {
            case .ok?:
                confirmation = Confirmation(
                    bookingId: nil,
                    userBookingStatus: .none,
                    waitlistPosition: nil
                )
                actionInFlight = false
            case .err(let err)?:
                actionError = err.error
                actionErrorMessage = "We couldn't complete that. Tap to try again."
                actionInFlight = false
            case .none:
                actionInFlight = false
            }
        }
    }

    func actionErrorShown() {
        actionError = nil
        actionErrorMessage = nil
    }


    func isWithinForfeitWindow() -> Bool {
        guard let startsAt = classInstance?.startsAt else { return false }
        return ForfeitPolicy.shared.isWithinForfeitWindow(startsAtIso: startsAt, now: Self.nowInstant())
    }


    func cancelTasks() {
        loadTask?.cancel()
        actionTask?.cancel()
    }


    private static func today() -> String {
        let now = Calendar.current.dateComponents([.year, .month, .day], from: Date())
        let year = now.year ?? 1970
        let month = now.month ?? 1
        let day = now.day ?? 1
        return String(format: "%04d-%02d-%02d", year, month, day)
    }

    private static func nowInstant() -> KotlinInstant {
        let millis = Int64((Date().timeIntervalSince1970 * 1000).rounded())
        return KotlinInstant.companion.fromEpochMilliseconds(epochMilliseconds: millis)
    }


    private static func toConfirmation(_ outcome: BookingOutcome) -> Confirmation {
        switch onEnum(of: outcome) {
        case .confirmed(let confirmed):
            return Confirmation(
                bookingId: confirmed.booking.bookingId,
                userBookingStatus: .booked,
                waitlistPosition: nil
            )
        case .waitlisted(let waitlisted):
            return Confirmation(
                bookingId: waitlisted.booking.bookingId,
                userBookingStatus: .waitlisted,
                waitlistPosition: waitlisted.position?.intValue
            )
        case .alreadyBooked(let already):
            return Confirmation(
                bookingId: nil,
                userBookingStatus: already.wasWaitlist ? .waitlisted : .booked,
                waitlistPosition: nil
            )
        }
    }
}
